cmd/snap-update-ns/utils.go

```diff
@@ -87,6 +89,49 @@ func IsReadOnly(dirFd int, dirName string) (bool, error) {
 	return false, nil
 }
 
+// IsTrustedTmpfs returns true if a directory belongs to a tmpfs mounted by snapd.
+//
+// The function inspects the directory (represented as both an open file
+// descriptor and the absolute path) and a list of changes that were applied to
+// the mount namespace. A directory is trusted if it is a tmpfs that was
+// mounted by snap-confine or snapd-update-ns. Note that sub-directories of a
+// trusted tmpfs are not considered trusted by this function.
+func IsTrustedTmpfs(dirFd int, dirName string, changes []*Change) (bool, error) {
+	var fsData syscall.Statfs_t
+	if err := sysFstatfs(dirFd, &fsData); err != nil {
+		return false, fmt.Errorf("cannot fstatfs %q: %s", dirName, err)
+	}
+	// If something is not a tmpfs it cannot be the trusted tmpfs we are looking for.
+	if fsData.Type != TmpfsMagic {
+		return false, nil
+	}
+	// Any of the past changes that mounted a tmpfs exactly at the directory we
+	// are inspecting is considered as trusted. This is conservative because it
+	// doesn't trust sub-directories of a trusted tmpfs. This approach is
+	// sufficient for the intended use.
+	//
+	// The algorithm goes over all the changes and picks up mount and unmount
+	// actions keeping track of what happened. The set of constraints in
+	// snap-update-ns and snapd prevent from mounting over an existing mount
+	// point so we don't need to consider e.g. a bind mount shadowing an active
+	// tmpfs.
```

Comment:
Perhaps a comment in the code that enforces this constraint that says 'If we change this constraint, please adjust isSnapdCreatedPrivateTmpfs() accordingly' (or similar) would be nice.

Reply:
I was thinking about this and there are some extra notes:

We don't have the code that walks mountinfo and refuses mounts over existing mount points. Such code existed in an older copy of snap-update-ns that was written in C.

We rename clashing mount entries (I added one extra test for that) makes most of what I said hold true... except for user mount entries that are currently technically allowed to shadow non-user mounts. The good thing is that we don't really have a way to define user mount entries anywhere except for that single document portal path.